[David Limantara] iP

xSaints19x opened the pull request:
# **Duke** ![flash (1)](https://user-images.githubusercontent.com/67327422/151771523-8464377a-668b-4861-af12-f59d1fde35fa.png)

> “Focus on being productive instead of busy.” - Tim Ferriss ([source](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwiD_fPcp9z1AhVTTmwGHbf8D7kQFnoECAcQAQ&url=https%3A%2F%2Fthestudentlifehacks.medium.com%2Ffocus-on-being-productive-instead-of-being-busy-tim-ferriss-aea9043da950&usg=AOvVaw0SQUEuyHY4oyQTk_g0H7gA))

Duke frees your mind of having to remember things you need to do. It's,
- Text-based
- Easy to learn
- ~~FAST~~ **_SUPER_ FAST** to use

Instead, you can have peace of mind and focus on **being productive**!😎

Tips on getting started:
1. Download it from [here](https://github.com/xSaints19x/ip/releases/tag/A-Jar)
2. Double-click it.
2. Add your tasks
3. Let it manage your tasks for you 😉

And it is absolutely **FREE**!

## Features of Duke
| Status | Features |
| :-: | :-: |
| ![checked (2)](https://user-images.githubusercontent.com/67327422/151773615-dc8db85a-0873-42d9-97af-cee7c722c65c.png)  | Managing Tasks  |
| ![coming-soon (5)](https://user-images.githubusercontent.com/67327422/151774085-2ad7d8d8-3ef4-4c0e-8f33-4b89b301aa69.png) | Managing Deadlines |
| ![coming-soon (5)](https://user-images.githubusercontent.com/67327422/151774085-2ad7d8d8-3ef4-4c0e-8f33-4b89b301aa69.png) | Managing Reminders |

If you are a Java programmer, you can use it to practice Java too. Here's the ```main``` method:
```Java
public class Main {
        public static void main(String[] args) {
                Application.launch(MainApp.class, args);
        }
}
```

Commit:
Add Gradle support

xSaints19x committed:
Level 1 Commit

xSaints19x committed:
Level 2 commit

xSaints19x committed:
Added a Mark as Done feature

xSaints19x committed:
Add support for tracking three types of tasks

xSaints19x committed:
Completed unit testing to ensure code correctness

xSaints19x committed:
Added error handlers for each sub-task

xSaints19x committed:
Added a delete feature to task list

xSaints19x committed:
Add a feature to save tasks

xSaints19x committed:
Add support for date formats of type yyyy-mm-dd

xSaints19x committed:
Merge branch 'branch-Level-8'

xSaints19x committed:
Refactor code to make it more OOP

xSaints19x committed:
Package the files into their respective packages

xSaints19x committed:
Add Junit tests to make sure code works as intended

xSaints19x committed:
Add Javadocs to all classes for greater readibility

xSaints19x committed:
Add a feature to find tasks in the task list that match the keyword given

xSaints19x committed:
Add output file

xSaints19x committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/Duke/util/Ui.java

xSaints19x committed:
Fix errors due to merging

xSaints19x committed:
Update README.md

xSaints19x committed:
Update code to fit coding standard

xSaints19x committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	README.md

xSaints19x committed:
Merge remote-tracking branch 'origin/add-gradle-support'

xSaints19x committed:
Add Gradle support to build Duke and run unit tests

xSaints19x committed:
Fix checkstyle issues

xSaints19x committed:
Revert "Fix checkstyle issues"

This reverts commit 1ffd6f4f78841fb98fef133c20bcf48273ce8332.

xSaints19x committed:
Fix Checkstyle issues

xSaints19x committed:
Add GUI to improve user experience

YiFei2 commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    /**
+     * Executes command by adding task into Duke.Duke.util.TaskList.
+     * @param taskList  List of tasks
```
I think there is supposed to be a blank line in between the description and your parameter section

YiFei2 commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,22 @@
+/**
+ * This Command class is the parent class for all commands to be executed.
+ */
+public abstract class Command {
```
I like the use of the abstract class Command and how inheritance is used for all the commands! 👍

YiFei2 commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,38 @@
+public class DeleteCommand extends Command {
+
+    public static final String COMMAND_WORD = "delete";
+    private final int taskIndex;
```
Did you intend for this to be final?

YiFei2 commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,38 @@
+public class FindCommand extends Command {
+
+    public static final String COMMAND_WORD = "find";
+    private final String keyword;
```
I think this should be in all uppercase if it is meant to be a constant 🙂

YiFei2 commented on src/main/java/duke/exception/DukeException.java:
```diff
@@ -0,0 +1,12 @@
+/**
+ * Print error messages
+ */
+public class DukeException extends Exception {
```
An interesting way to use DukeException class that I've seen from another PR was to have different types of DukeExceptions extending from this class with their own error messages. The programs will then throw these exceptions instead of the various java exceptions. Just sharing something I learnt! You could consider doing it too if you think that it's better 😄

YiFei2 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    public String getStatusIcon() {
+        return (isDone ? "X" : " "); // mark done task with X
```
```suggestion
        // mark done task with X
        return (isDone ? "X" : " "); 
```
I'm not sure but maybe the comment is supposed to be like this?

YiFei2 reviewed (commented):
Looks good!

Oddcorner commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -0,0 +1,25 @@
+
+    /**
+     * Executes command by printing exit message.
+     * @param taskList List of tasks.
```
Should there be a line break between the description and the block tags here? I have also noticed other places where this is omitted.

Oddcorner commented on src/main/java/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,38 @@
+    /**
+     * Constructor for DeleteCommand which provides an index to delete.
+     *
+     * @param taskIndex 0-based index.
```
I like how you specified the index being 0-based here. It's always confusing switching between 0-based and 1-based! 👍

Oddcorner commented on src/main/java/duke/gui/MainWindow.java:
```diff
@@ -0,0 +1,92 @@
+import duke.Duke;
+import duke.exception.DukeException;
+import duke.util.Ui;
+import javafx.fxml.FXML;
```
I think there should be an empty line between imports from duke package and JavaFX package to be consistent with the way you have been doing your imports in other files.

Oddcorner commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,81 @@
+
+    /**
+     * Save all tasks into a file
+     */
```
Should there be a @param tag here as this method has an `ArrayList<Task>` parameter?

Oddcorner commented on src/main/java/duke/util/TaskList.java:
```diff
@@ -0,0 +1,205 @@
+            Task task = taskList.get(taskIndex);
+            this.taskList.remove(taskIndex);
+            this.taskCount--;
+            str.append(ui.deleteTask() + task+ "\n" + ui.requestNextCommand());
```
There is a missing white space here.
```suggestion
            str.append(ui.deleteTask() + task + "\n" + ui.requestNextCommand());
```

Oddcorner commented on src/main/java/duke/util/TaskList.java:
```diff
@@ -0,0 +1,205 @@
+    /**
+     * Mark a task in list of tasks as done
+     *
+     * @param taskIndex 0-based index of task number.
```
Should there be a @return tag here as this method returns a String?

Oddcorner reviewed (commented):
Overall you followed the coding standard really well! 👍 

Just little things to be aware of are some missing line breaks and tags like `@return` and `@params` in various javadocs comments throughout the code.

xSaints19x committed:
Update GUI to have more images to look more vibrant

xSaints19x committed:
Add assert statements

Modify code for better readibility

Some methods' Javadoc comments are too close together and missing
some tags.

Javadoc comments being too close together made it difficult to read
description and missing tags cause users not to be aware of what the
method with the missing return tag, return value is.

Adding whitespaces between Javadoc method description and params
tag to help with better code readibility. Add missing return tags
for methods to ensure users know the what each method returns.

xSaints19x committed:
Merge pull request #1 from xSaints19x/branch-A-Assertions

xSaints19x committed:
Consolidate Duplicate Conditional Fragments

The same fragment of code is in all branches of a conditional
expression.

Move it outside of the expression

This is to reduce code duplication.

xSaints19x committed:
Merge branch 'master' into branch-A-CodeQuality

xSaints19x committed:
Merge pull request #2 from xSaints19x/branch-A-CodeQuality

Consolidate Duplicate Conditional Fragments

xSaints19x committed:
Fix some minor code redundancy and issues

xSaints19x committed:
Add Continuous Integration (CI) to repo

xSaints19x committed:
Add help command for users to display the list of commands that the
they can type in and the format in which they should type it in

xSaints19x committed:
Add more images and gifs to better interact with users

xSaints19x committed:
Extract methods in Parser Class to make code easier to understand

xSaints19x committed:
Improve GUI to make it more interactive and made minor edits to
some method's Javadocs description

xSaints19x committed:
Update README.md

xSaints19x committed:
Add final touch-ups to the program before release
